pkg/controller/stack/stack_controller.go

```diff
@@ -199,18 +255,33 @@ func (r *ReconcileStack) Reconcile(ctx context.Context, request reconcile.Reques
 		return reconcile.Result{}, err
 	}
 
-	// This makes sure the status reflects the outcome of reconcilation. Any non-error return means
-	// the object definition was observed, whether the object ended up in a ready state or not. An
-	// error return (now we have successfully fetched the object) means it is "in progress" and not
-	// ready.
+	// This makes sure the status reflects the outcome of reconcilation. An error return can either
+	// indicate a retry is needed (RetryError), processing has stalled (StallError), or a crash
+	// (anything else).
 	saveStatus := func() {
-		if reterr == nil {
-			instance.Status.ObservedGeneration = instance.GetGeneration()
-		} else {
-			// controller-runtime will requeue the object, so reflect that in the conditions by
-			// saying it is still in progress.
-			instance.Status.MarkReconcilingCondition(pulumiv1.ReconcilingRetryReason, reterr.Error())
+		if reterr != nil {
```

Comment:
This error handling appears to introduce new semantics around how/when retires and stalls are triggered. Is this strictly related to the namespace isolation changes?  Are we confident these changes are well covered in tests?

Reply:
Re test coverage: #316 included tests that _some_ arbitrary stalling problem resulted in status conditions indicating a stall, and _some_ arbitrary retryable problem resulted in a status conditions indicating "in progress". This PR has tests that with namespace isolation, using a cross-namespace ref results in a stall, and with namespace isolation waived, it doesn't (all else going well).

For _specific_ stalling or retryable problems -- I'll leave those to the other, hypothetical PR. They could have been individually tested here, I was relying on natural induction :-)

Reply:
The more recent attempt uses the same error type protocol, but only for cross-namespace refs causing a stall, and only at the site of calling the helper; so, everything has the same retry/stall semantics.